HDDS-11938. [Website v2] [Docs] [Quick Start] Try Ozone With Kubernetes

Eric977 opened the pull request:
## What changes were proposed in this pull request?

[HDDS-11938](https://issues.apache.org/jira/browse/HDDS-11938). [Website v2] [Docs] [Quick Start] Try Ozone With Kubernetes

Please describe your PR in detail:
Migrate doc from https://ozone.apache.org/docs/edge/start/kubernetes.html to https://ozone-site-v2.staged.apache.org/docs/quick-start/installation/kubernetes

## What is the link to the Apache Jira?
https://issues.apache.org/jira/browse/HDDS-11938

## How was this patch tested?

Previewed locally.

![image](https://github.com/user-attachments/assets/abe87bcb-1c43-4f85-bf9c-f7f8d604d0b1)

adoroszlai commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,52 @@
+* kubectl
+
+As the *apache/ozone* Docker images are available from the dockerhub the deployment process is very similar to Minikube deployment. The only big difference is that we have dedicated set of k8s files for hosted clusters (for example we can use one Datanode per host)
+Deploy to kubernetes
```
This is supposed to be a heading (was wrong in the existing docs).

```suggestion
### Deploy to Kubernetes
```

(Suggesting 3rd level heading, assuming Minikube and "hosted cluster" content would be 2nd level, and this is under "hosted cluster".)

adoroszlai reviewed (commented):
Thanks @Eric977 for working on this.

In the old docs, Kubernetes topic is split to two pages:
- [Minikube & Ozone](https://ozone.apache.org/docs/edge/start/minikube.html)
- [Ozone on Kubernetes](https://ozone.apache.org/docs/edge/start/kubernetes.html)

I think we should import both to this single page.  The PR is now importing only the second item, but the text references the first one ("Minikube deployment").  As a reader, I wouldn't know what that is.  Not sure what would be the best heading for current content.  (Does "Hosted Cluster" make sense?)

I also think we need to improve the text in follow-up task(s).  As a start, it's best to import it as-is.  However, we should make some trivial fixes, like changing "the dockerhub" to "Docker Hub".  This makes it unnecessary to define "dockerhub" as a word (it is not).

Eric977 commented:
Thank you for your feedback! I've addressed the missing pages and added them to the migration. 

Regarding the heading, I’ve applied the suggested change to 'Hosted Cluster.' Do you think 'Hosted Kubernetes Cluster' might be clearer?

adoroszlai commented:
Thanks for the updates.

> Do you think 'Hosted Kubernetes Cluster' might be clearer?

Yes, that may be better.

errose28 commented:
> I also think we need to improve the text in follow-up task(s). As a start, it's best to import it as-is. 

I would like avoid importing unfinished content into the new website, it will be hard to track and we will just accumulate debt on an already new project. Remember we are going to have 200-300 pages worth of docs. As part of testing for this PR, please try out the steps locally to verify that they are clear and working.

adoroszlai commented on docs/02-quick-start/01-installation/02-kubernetes.md:
````diff
@@ -1,7 +1,106 @@
+
+``` bash
+kubectl get pod
+Access the services
+```
+
````
This is also supposed to be a heading:

```
### Access the services
```

Eric977 replied:
If I add the heading, it would result in the MD024/no-duplicate-heading error at /root/project/ozone-site/docs/02-quick-start/01-installation/02-kubernetes.md:100 under '### Access the services'. Should I simply drop the heading?

adoroszlai replied:
Hmm, good point.

Possible solutions:
- split the page, as in the existing website
- come up with unique headings
- remove these subsection headings
- is there some flag to allow duplicate headings?

@errose28 what's your take?

errose28 replied:
Ah this was a miss when I set up the CI. We only want the duplicate header check for siblings, which is not the case here. See https://github.com/DavidAnson/markdownlint/blob/v0.35.0/doc/md024.md and please set `siblings_only=true` in `markdownlint.yaml` as part of this PR to fix this issue. Then these headings should be allowed.

Eric977 commented:
the command  
`kubectl apply -f .`
 result in the following error:
`error: error validating "kustomization.yaml": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false`

To resolve this issue, replace the command with:
`kubectl apply -k .`

The issue and its solution are discussed in detail in [this GitHub issue](https://github.com/kubernetes-sigs/kustomize/issues/738).

Currently, I am testing the documentation in my local environment (wsl on window) and have encountered some additional issues. It seems that the latest release (1.4.1) has no corresponding Docker image, which results in pod initialization failures.

After switching to version 1.4.0,  I encounter the following error and am still investigating:
```root@DESKTOP-UHVDC0I:~/project/ozone-1.4.0/kubernetes/examples/ozone# sudo kubectl get pod
NAME         READY   STATUS       RESTARTS        AGE
datanode-0   0/1     Error        8 (2m8s ago)    13m
datanode-1   0/1     Pending      0               16m
om-0         1/1     Running      8 (5m44s ago)   17m
s3g-0        1/1     Running      0               17m
scm-0        0/1     Init:Error   8 (2m49s ago)   13m```

adoroszlai commented:
> It seems that the latest release (1.4.1) has no corresponding Docker image

Thank you for spotting this.  Recently we switched to Rocky Linux, and published new images for existing versions with tags like `1.4.0-rocky` to avoid disruption for users of existing images.  1.4.1 was the first release to only have `-rocky` image.  But the kubernetes examples were not updated, so they still point to `1.4.1`.  I think we should publish additional tag `1.4.1` for the `1.4.1-rocky` image.

adoroszlai commented:
Docker Image `apache/ozone:1.4.1` is now available.

Eric977 commented:
> Docker Image apache/ozone:1.4.1 is now available.
Thanks for your help

> After switching to version 1.4.0, I encounter the following error and am still investigating:

The issue was due to a misconfiguration in my Kubernetes storage. After fixing it, everything started running correctly.
So I believe the content of the document is valid.

adoroszlai commented:
Thanks @Eric977 for testing and updating the doc.  @errose28 please take another look.

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,107 @@
+**Requirements:**
+
+* Working minikube setup
+* kubectl
```
Please use `-` for unordered lists since it is consistent with the existing pages in the new site. This was supposed to be enforced by markdownlint but I made a mistake in the config. I'll push a fix for the markdownlint config soon.

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,107 @@
+* Working minikube setup
+* kubectl
+
+`kubernetes/examples` folder of the Ozone distribution contains Kubernetes deployment resource files for multiple use cases. By default the Kubernetes resource files are configured to use *apache/ozone* image from the Docker Hub.
```
```suggestion
The `kubernetes/examples` directory of the Ozone distribution contains Kubernetes deployment resource files for multiple use cases. By default the Kubernetes resource files are configured to use the [apache/ozone](https://hub.docker.com/r/apache/ozone) image from Docker Hub.
```

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
````diff
@@ -1,7 +1,107 @@
+kubectl get pod
+```
+
+Note: the `kubernetes/examples/minikube` resource set is optimized for minikube usage:
````
For callouts like this we should docusaurus admonitions as described in the [contributing guide](https://github.com/apache/ozone-site/blob/HDDS-9225-website-v2/CONTRIBUTING.md#page-format).

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,107 @@
+Note: the `kubernetes/examples/minikube` resource set is optimized for minikube usage:
+
+* You can have multiple Datanodes even if you have only one host (in a real production cluster usually you need one Datanode per physical host)
+* The services are published with node port
```
Can you elaborate on what this means?

Eric977 replied:
Node port is an opening port that allowing externel traffic to access the service.
The detailed explaination could be found [here](https://kubernetes.io/docs/concepts/services-networking/service/)

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
````diff
@@ -1,7 +1,107 @@
+
+``` bash
+minikube service s3g-public
+Opening kubernetes service default/s3g-public in default browser...
````
I haven't tried this, but what does this display? Is it the s3g web UI?

Eric977 replied:
Yes, it will provide a URL to access a website.
![image](https://github.com/user-attachments/assets/a3137146-07e1-40d4-a9bb-3b6baa3fa35d)

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
````diff
@@ -1,7 +1,107 @@
+scm-public   NodePort    10.105.231.28   <none>        9876:32171/TCP   27s
+```
+
+Minikube contains a convenience command to access any of the NodePort services:
````
```suggestion
Minikube provides a convenience command to access any of the NodePort services:
```

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,107 @@
+
+### Access the services
+
+Now you can access any of the services. For each web endpoint an additional NodeType service is defined in the minikube k8s resource set. NodeType services are available via a generated port of any of the host nodes:
```
```suggestion
Now you can access any of the services. For each web endpoint an additional NodeType service is defined in the minikube k8s resource set. NodeType services are available via a generated port on any of the host nodes:
```

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,107 @@
+
+* Working minikube setup
+* kubectl
+
```
```suggestion
### Deploy the Services
```

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
````diff
@@ -1,7 +1,107 @@
+Opening kubernetes service default/s3g-public in default browser...
+```
+
+## Hosted Kubernetes Cluster
````
The formatting suggestions above also apply to this section

errose28 commented on docs/02-quick-start/01-installation/02-kubernetes.md:
````diff
@@ -1,7 +1,107 @@
+``` bash
+kubectl port-forward s3g-0 9878:9878
+kubectl port-forward scm-0 9876:9876
+```
````
At some point here the we should run `ozone version` in a pod as part of the test that things were set up correctly. More thorough testing will be done in the following "Reading and Writing Data" section.

errose28 reviewed (commented):
Thanks for adding this @Eric977

Eric977 commented:
Hi @errose28 

Thanks for the review, and sorry for the mistakes. I have fixed the comments above.

adoroszlai commented on docs/02-quick-start/01-installation/02-kubernetes.md:
```diff
@@ -1,7 +1,107 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
```
I think license is included in the original docs in `apache/ozone` since they are part of the official source release.  But the website is not "released", so we can remove the license header.
